distribution/xfer/download.go

```diff
@@ -62,6 +63,9 @@ type DownloadDescriptor interface {
 	// if it is unknown (for example, if it has not been downloaded
 	// before).
 	DiffID() (layer.DiffID, error)
+	// ForeignSource returns nil or the foreign layer's information and
+	// source.
+	ForeignSource() *distribution.Descriptor
```

Comment:
We already store digests for all layers. See https://github.com/docker/docker/blob/master/distribution/metadata/v2_metadata_service.go